internal/provider/resource_gitlab_user_gpgkey.go

```diff
@@ -0,0 +1,187 @@
+				Type:        schema.TypeInt,
+				ForceNew:    true,
+				Optional:    true,
+			},
+			"key": {
+				Description: "The armored GPG public key.",
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Required:    true,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					strippedOld := strings.TrimSpace(old)
+					strippedNew := strings.TrimSpace(new)
+					return strippedOld == strippedNew
+				},
+			},
+			"key_id": {
+				Description: "The ID of the GPG key.",
+				Type:        schema.TypeInt,
+				Computed:    true,
+			},
+			"created_at": {
+				Description: "The time when this key was created in GitLab.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+		},
+	}
+})
+
+func resourceGitlabUserGPGKeyCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+
+	options := &gitlab.AddGPGKeyOptions{
+		Key:   gitlab.String(strings.TrimSpace(d.Get("key").(string))),
+	}
+
+	var key *gitlab.GPGKey
+	var err error
+	userID, userIDOk := d.GetOk("user_id")
+	if userIDOk {
+		isAdmin, isAdminErr := isCurrentUserAdmin(ctx, client)
+		if isAdminErr != nil {
+			return diag.Errorf("failed to check if user is admin for configuring GPG keys for a user")
+		}
+		if !isAdmin {
+			return diag.Errorf("current user needs to be admin for configuring GPG keys for a user")
+		}
+		key, _, err = client.Users.AddGPGKeyForUser(userID.(int), options, gitlab.WithContext(ctx))
+	} else {
+		key, _, err = client.Users.AddGPGKey(options, gitlab.WithContext(ctx))
+	}
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	keyIDForID := fmt.Sprintf("%d", key.ID)
+	if userIDOk {
+		userIDForID := fmt.Sprintf("%d", userID)
+		d.SetId(buildTwoPartID(&userIDForID, &keyIDForID))
+	} else {
+		d.SetId(keyIDForID)
+	}
+	return resourceGitlabUserGPGKeyRead(ctx, d, meta)
+}
+
+func resourceGitlabUserGPGKeyRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+
+	userID, keyID, err := resourceGitlabUserGPGKeyParseID(d.Id())
+	if err != nil {
+		return diag.Errorf("unable to parse user GPG key resource id: %s: %v", d.Id(), err)
+	}
+
+	var keys []*gitlab.GPGKey
+	if userID != 0 {
+		keys, _, err = client.Users.ListGPGKeysForUser(userID, gitlab.WithContext(ctx))
+	} else {
+		keys, _, err = client.Users.ListGPGKeys(gitlab.WithContext(ctx))
+	}
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	var key *gitlab.GPGKey
+	for _, k := range keys {
+		if k.ID == keyID {
+			key = k
+			break
+		}
+	}
+
+	if key == nil {
+		log.Printf("Could not find GPG key %d for user %d", keyID, userID)
+		d.SetId("")
+		return nil
+	}
+
+	if userID != 0 {
+		d.Set("user_id", userID)
+	}
+	d.Set("key_id", keyID)
+	d.Set("key", strings.TrimSpace(key.Key))
+	if key.CreatedAt != nil {
```

Comment:
NIT: probably always could to check, but did you actually run into this? I don't think it will every be `nil` ...

Reply:
> did you actually run into this?

Nope. Again, I copied codes from `resource_gitlab_user_sshkey.go` without checking it in depth lol